### DISCUSSION TITLE
Monitored item refresh rate

### DISCUSSION THREAD

**Discussion by EmmanuelP** [locka99/opcua#163] [Q&A] [answered]
Hi,

Thanks for your work on this rust opcua implementation ! :heart: 

I'm currently investigating the use of rust opcua as a replacement of our distributed process communication protocol.

One of the test consist in a client retrieving the value of a node use the MonitoredItem service. The node is a simple integer counter incremented every 100ms via `add_polling_action` (the code is based on rust opcua simple-server).

The problem is either in a simple client, or in UAExpert, I'm not able to get all the values. Sometimes there is several seconds between 2 updates in the client, even if the publishing interval is set to 100ms and the queue is 10 items long.

Is there a setting I have overlooked ?

**Comment by schroeder-** [comment]
There are two parameters, publishing interval and sampling interval.
Publishing interval is the time where all changed values are send to the client. 
Sampling interval is the time the variables are read for subscription.

On the server side you can change the value via toml file:
```
limits:
  min_sampling_interval: 0.05
  min_publishing_interval: 0.05
```
or via 
ServerConfig -> limits.min_sampling_interval

On the client side you can test around with UAExpert Data Logger View.
![image](https://user-images.githubusercontent.com/1273483/153906106-44be0805-6d9a-46ca-a79e-5290d51ccfe7.png)

**Comment by EmmanuelP** [comment]
Thanks @schroeder- .

My server is built using the following code:

``` rs
    let mut server: Server = ServerBuilder::new()
        .application_name("producer")
        .application_uri("urn:producer")
        .product_uri("urn:producer")
        .discovery_server_url(Some(constants::DEFAULT_DISCOVERY_SERVER_URL.to_string()))
        .endpoints(vec![("none", ServerEndpoint::new_none(path, &user_token_ids))])
        .discovery_urls(vec![path.into()])
        .server().unwrap();
```

Is it possible to set the intervals here ? Or do I need to resort to a configuration file or build Server from a ServerConfig struct ?

**Comment by schroeder-** [comment, marked as answer]
Try this:
```
let mut cfg = ServerBuilder::new()
        .application_name("producer")
        .application_uri("urn:producer")
        .product_uri("urn:producer")
       ......
        .config();
cfg.limits.min_sampling_interval = 0.1;
cfg.limits.min_publishing_interval = 0.1;
let mut server = Server::new(cfg);
```

**Comment by EmmanuelP** [comment]
It was pretty obvious in the end. I'm learning opcua the spec, opcua the rust implementation and rust itself, my brain is overheating.

Thanks @schroeder-
